File: 300-longest-increasing-subsequence/longest-increasing-subsequence.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var lengthOfLIS = function(nums) {
    // return lenLISRecursive(nums, 0, -1);
    const dp = new Array(nums.length).fill().map(() => new Array(nums.length + 1).fill(-1));
    return lenLISMemo(nums, 0, -1, dp);
};

function lenLISMemo(nums, i, prevIdx, dp) {
    if(i === nums.length) {
        return 0;
    }

    if(dp[i][prevIdx+1] !== -1) {
        return dp[i][prevIdx+1];
    }

    // NOT TAKE
    const notTake = 0 + lenLISMemo(nums, i+1, prevIdx, dp);

    // TAKE
    let take = 0;
    if(prevIdx == -1 || nums[i] > nums[prevIdx]) {
        take = 1 + lenLISMemo(nums, i+1, i, dp);
    }

    return dp[i][prevIdx+1] = Math.max(take, notTake);
}

function lenLISRecursive(nums, i, prevIdx) {
    if(i === nums.length) {
        return 0;
    }

    // NOT TAKE
    const notTake = 0 + lenLISRecursive(nums, i+1, prevIdx);

    // TAKE
    let take = 0;
    if(prevIdx == -1 || nums[i] > nums[prevIdx]) {
        take = 1 + lenLISRecursive(nums, i+1, i);
    }

    return Math.max(take, notTake);
}

File: 300-longest-increasing-subsequence/longest-increasing-subsequence.cpp
class Solution {
public:
    int lengthOfLIS(vector<int>& nums) {
        // return recursive(nums, 0, -1);
        int n = nums.size();
        vector<vector<int>> dp(n, vector<int>(n + 1, -1));
        return memoized(nums, 0, -1, dp);
    }

    int memoized(vector<int>& nums, int i, int prevIdx,
                 vector<vector<int>>& dp) {
        int n = nums.size();
        if (i == n) {
            return 0;
        }

        if (dp[i][prevIdx + 1] != -1) {
            return dp[i][prevIdx + 1];
        }

        // not take
        int len = 0 + memoized(nums, i + 1, prevIdx, dp);
        // take - ONLY WHEN prevVal < currVal
        if (prevIdx == -1 || nums[prevIdx] < nums[i]) {
            // YEH ELEM LIYA HAI ==> prevIdx = i
            len = max(len, 1 + memoized(nums, i + 1, i, dp));
        }

        return dp[i][prevIdx + 1] = len;
    }

    int recursive(vector<int>& nums, int i, int prevIdx) {
        int n = nums.size();
        if (i == n) {
            return 0;
        }

        // not take
        int len = 0 + recursive(nums, i + 1, prevIdx);
        // take - ONLY WHEN prevVal < currVal
        if (prevIdx == -1 || nums[prevIdx] < nums[i]) {
            // YEH ELEM LIYA HAI ==> prevIdx = i
            len = max(len, 1 + recursive(nums, i + 1, i));
        }

        return len;
    }
};